internal/oci/ecr/credentials_store.go

```diff
@@ -0,0 +1,87 @@
+package ecr
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"strings"
+	"sync"
+	"time"
+
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"oras.land/oras-go/v2/registry/remote/credentials"
+)
+
+func defaultClientFunc(serverAddress string) Client {
+	switch {
+	case strings.HasPrefix(serverAddress, "public.ecr.aws"):
+		return &publicClient{}
+	default:
+		return &privateClient{}
+	}
+}
+
+func NewCredentialsStore() credentials.Store {
+	return &CredentialsStore{
+		cache:      map[string]cacheItem{},
+		clientFunc: defaultClientFunc,
+	}
+}
+
+type cacheItem struct {
+	credential auth.Credential
+	expiresAt  time.Time
+}
+
+type CredentialsStore struct {
+	mu         sync.Mutex
+	cache      map[string]cacheItem
+	clientFunc func(serverAddress string) Client
+}
+
+// Get retrieves credentials from the store for the given server address.
+func (s *CredentialsStore) Get(ctx context.Context, serverAddress string) (auth.Credential, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if item, ok := s.cache[serverAddress]; ok && time.Now().UTC().Before(item.expiresAt) {
+		return item.credential, nil
+	}
+
+	token, expiresAt, err := s.clientFunc(serverAddress).GetAuthorizationToken(ctx)
+	if err != nil {
+		return auth.EmptyCredential, err
+	}
+	credential, err := s.extractCredential(token)
+	if err != nil {
+		return auth.EmptyCredential, err
+	}
+	s.cache[serverAddress] = cacheItem{credential, expiresAt}
+	return credential, nil
+}
+
+// Put saves credentials into the store for the given server address.
+func (s *CredentialsStore) Put(ctx context.Context, serverAddress string, cred auth.Credential) error {
+	return errors.ErrUnsupported
```

Comment:
I'm curious about this, does oras treat this error correctly? Like, if we return this error it'll just ignore?

Reply:
I used `credentials.Credential()` from oras which requires `Store` interface.  In reality that function only uses `Get` from store. Probably I should clean up it more.